broker/cloud_functions/lsst/ps_to_gcs/main.py

```diff
@@ -0,0 +1,198 @@
+            "format": "avro",
+        }
+    )
+
+    # get bucket based on the alert's schema version and store the Avro file
+    bucket = BUCKETS.get(schema_version)
+    blob = bucket.blob(filename)
+    blob.metadata = create_file_metadata(alert_dict, context)
+
+    # raise a PreconditionFailed exception if filename already exists in the bucket using "if_generation_match=0"
+    # let it raise. the main function will catch it and then drop the message.
+    blob.upload_from_file(io.BytesIO(alert_bytes), if_generation_match=0)
+
+    # Cloud Storage says this is not a duplicate, so now we publish the broker's main "alerts" stream
+    return publish_outgoing_alert(
+        topic_name=ALERTS_TOPIC.name,
+        message=alert_bytes,
+        attributes={
+            "objectId": str(alert_dict["diaObject"]["diaObjectId"]),
+            "sourceId": str(alert_dict["diaSource"]["diaSourceId"]),
+            "schema_version": schema_version,
+            **attributes,
+        },
+    )
+
+
+def deserialize_confluent_wire_header(raw):
+    """Parses the byte prefix for Confluent Wire Format-style Kafka messages.
+    Parameters
+    ----------
+    raw : `bytes`
+        The 5-byte encoded message prefix.
+    Returns
+    -------
+    schema_version : `int`
+        A version number which indicates the Confluent Schema Registry ID
+        number of the Avro schema used to encode the message that follows this
+        header.
+    """
+    _, version = _ConfluentWireFormatHeader.unpack(raw)
+
+    return version
+
+
+def generate_alert_filename(aname: dict) -> str:
+    """
+    Generate the filename of an alert stored to a Cloud Storage bucket.
+
+    Args:
+        aname:
+            Components to create the filename. Required key/value pairs are those needed to create a parsed filename.
+            Extra keys are ignored.
+
+    Returns:
+        str: The formatted filename as "{topic}/{objectId}/{sourceId}.{format}".
+    """
+    topic = aname.get("topic", "no_topic")
+    object_id = aname.get("objectId")
+    source_id = aname.get("sourceId")
+    file_format = aname.get("format", "avro")
+
+    return f"{topic}/{object_id}/{source_id}.{file_format}"
+
+
+def create_file_metadata(alert_dict: dict, context):
+    """Return key/value pairs to be attached to the file as metadata."""
+    metadata = {"file_origin_message_id": context.event_id}
+    metadata["diaObjectId"] = alert_dict["diaObject"]["diaObjectId"]
+    metadata["diaSourceId"] = alert_dict["diaSource"]["diaSourceId"]
+    metadata["ra"] = alert_dict["diaSource"]["ra"]
+    metadata["dec"] = alert_dict["diaSource"]["dec"]
+    return metadata
+
+
+def publish_outgoing_alert(
+    topic_name: str, message: bytes, attributes: Optional[dict] = None
+) -> str:
+    """Publish messages to a Pub/Sub topic."""
+
+    # enforce bytes type for message
+    if not isinstance(message, bytes):
+        raise TypeError("`message` must be bytes or a dict.")
```

Comment:
In this case (publishing our main `lsst-alerts` stream), we want this to be the original, Avro-serialized bytes. We don't want to publish a dict here.